File: sorting/merge_sort/MergeSort.java
import java.util.ArrayList;
import java.util.List;

class MergeSort {
	public static <T extends Comparable<T>> List<T> sort(List<T> elements) {
		int size = elements.size();

		if (size <= 1) {
			return elements;
		}

		int midpoint = size / 2;

		List<T> left = sort(elements.subList(0, midpoint));
		List<T> right = sort(elements.subList(midpoint, size));

		// TODO(durandj): this could be done with iterators
		int leftIndex = 0;
		int rightIndex = 0;
		List<T> sorted = new ArrayList<T>();

		while (leftIndex < midpoint && rightIndex < size - midpoint) {
			T leftValue = left.get(leftIndex);
			T rightValue = right.get(rightIndex);

			if (leftValue.compareTo(rightValue) < 0) {
				sorted.add(leftValue);
				leftIndex++;
			}
			else {
				sorted.add(rightValue);
				rightIndex++;
			}
		}

		while (leftIndex < midpoint) {
			sorted.add(left.get(leftIndex));
			leftIndex++;
		}

		while (rightIndex < size - midpoint) {
			sorted.add(right.get(rightIndex));
			rightIndex++;
		}

		return sorted;
	}
}

File: sorting/merge_sort/merge_sort.js
'use strict';

function swap(elements, index0, index1) {
	const tmp = elements[index0];
	elements[index0] = elements[index1];
	elements[index1] = tmp;
}

function mergeSort(elements) {
	const size = elements.length;

	if (size <= 1) {
		return elements;
	}

	const midpoint = Math.floor(size / 2);
	const left = mergeSort(elements.slice(0, midpoint));
	const right = mergeSort(elements.slice(midpoint, size));
	let sorted = [];

	while (left.length > 0 && right.length > 0) {
		if (left[0] <= right[0]) {
			sorted.push(left.shift());
		}
		else {
			sorted.push(right.shift());
		}
	}

	return sorted.concat(left).concat(right);
}

